sfs/time/drivingfunction.py

```diff
@@ -255,3 +256,194 @@ def apply_delays(signal, delays):
+        if normalize:
+            k = _normalize_digital_filter_gain(s0, sinf, z0, zinf, fs=fs)
+        else:
+            k = 1
+        sos.append(zpk2sos(z0, zinf, k, pairing='nearest'))
+        # TODO: normalize the SOS filters individually?
+    phaseshift = phipw + np.pi - phi0
+    return delay, weight, sos, phaseshift
+
+
+def nfchoa_25d_point(x0, r0, xs, max_order=None, c=None, fs=44100, normalize=True):
+    """Point source by 2.5-dimensional NFC-HOA.
+
+    Parameters
+    ----------
+    x0 : (N, 3) array_like
+        Sequence of secondary source positions.
+    r0 : float
+        Radius of the circular secondary source distribution
+    xs : (3,) array_like
+        Virtual source position.
+    max_order : int
+        Ambisonics order
+    c : float, optional
+        Speed of sound
+
+    Returns
+    -------
+    delay : float
+        Overall delay (common to all secondary source)
+    weight : float
+        Overall weight (common to all secondary sources)
+    sos : list
+        Second-order section filters
+    phaseshift : float
+        Phase shift
+
+    References
+    ----------
+        S. Spors, V. Kuscher, J. Ahrens (2011) - "Efficient realization of
+       model-based rendering for 2.5-dimensional near-field compensated higher
+       order Ambisonics", WASPAA, p. 61-64
+
+       See Eq.(11)
+   """
+    max_order = util.max_order_circular_harmonics(len(x0), max_order)
+    if c is None:
+        c = defs.c
+
+    x0 = util.asarray_of_rows(x0)
+    xs = util.asarray_1d(xs)
+    phi0, _, _ = util.cart2sph(*x0.T)
+    phi, _, r = util.cart2sph(*xs)
+
+    delay = (r-r0)/c
+    weight = 1/2/np.pi/r
+    sos = []
+    for m in range(0, max_order+1):
+        _, p, k = besselap(m, norm='delay')
+        # TODO: modify "besselap" for very high orders (>150)
+        s0 = (c/r)*p
+        sinf = (c/r0)*p
+        z0 = np.exp(s0/fs)
+        zinf = np.exp(sinf/fs)
+        # TODO: select matched-z or bilinear transform
+        if normalize:
+            k = _normalize_digital_filter_gain(s0, sinf, z0, zinf, fs=fs)
+        else:
+            k = 1
+        sos.append(zpk2sos(z0, zinf, k, pairing='nearest'))
+        # TODO: normalize the SOS filters individually?
+    phaseshift = phi0 - phi
+    return delay, weight, sos, phaseshift
+
+
+def nfchoa_driving_signals(delay, weight, sos, phaseshift, signal, max_order=None, fs=44100):
+    """Get NFC-HOA driving signals per secondary source.
+
+    Parameters
+    ----------
+    delay : float
+        Overall delay (common to all secondary source)
+    weight : float
+        Overall weight (common to all secondary sources)
+    sos : dictionary
+        Second-order section filters
+    phaseshift : (C,) array_like
+        Phase shift
+    signal : (N,) array_like
+        Excitation signal (mono) which gets weighted and delayed.
+    fs: int, optional
+        Sampling frequency in Hertz.
+
+    Returns
+    -------
```

Comment:
I think it would make sense to return a `DelayedSignal`, like in `driving_signals()`.